android/app/src/main/java/com/dismoi/scout/floating/layout/Message.kt

```diff
@@ -1,20 +1,157 @@
+    set (notices) {
+      val carouselView = findViewById<CarouselView>(R.id.carouselView)
+      carouselView.pageCount = notices.size
+      field = notices
+    }
+
+  private var inflater: LayoutInflater = context.getSystemService(Service.LAYOUT_INFLATER_SERVICE) as LayoutInflater
+
+  // TODO Fix fonts usage
+  //    val type = Typeface.createFromAsset(context.assets, "fonts/Helvetica.ttf")
+  //    val typeBold = Typeface.createFromAsset(context.assets, "fonts/Helvetica-Bold.ttf")
+
+  private var carouselListener = ViewListener { position ->
+    val liveNotice = notices[position]
+
+    // TODO Extract in fun
+    val noticeView = inflater.inflate(R.layout.message, null)
+
+    val noticeObserver = Observer<JSONObject> { noticeContent ->
+      if (liveNotice.value == null) {
+        return@Observer
+      }
+
+      val message = noticeContent.getString("message")
+      val contributor = noticeContent.getJSONObject("contributor")
+      val contributorName = contributor.getString("name") // TODO don’t crash if no name
+      val contributorAvatarUrl =
+        if (contributor.has("avatar") && !contributor.isNull("avatar"))
+          URL(
+            contributor.getJSONObject("avatar").getJSONObject("normal").getString("url")
+          )
+        else null
+
+      val modified = noticeContent.getString("modified") // TODO should be parsed
+
+      val messageView = noticeView.findViewById<TextView>(R.id.link)
+      messageView.text = message.toSpanned()
+//    messageView.typeface = type
+      messageView.handleUrlClicks { urlLinkToClick ->
+        Log.d("Message", "Clicked link")
+        // TODO Legacy : sendEventToReactNative("URL_CLICK_LINK", Uri.parse(urlLinkToClick).toString())
+      }
+
+      val contributorNameView = noticeView.findViewById<TextView>(R.id.name)
+      contributorNameView.text = contributorName
+//    contributorNameView.typeface = typeBold // TODO Should be defined in view, right ?
+
+      val modifiedDateView = noticeView.findViewById<TextView>(R.id.date)
+      modifiedDateView.text = modified // TODO should be formatted correctly
+//    modifiedDateView.typeface = typeBold // TODO Should be defined in view, right ?
+
+      val deleteButton = noticeView.findViewById<ImageView>(R.id.closeNotice)
+      deleteButton.setOnClickListener {
+        // TODO What to do onClick
+      }
+
+      if (contributorAvatarUrl != null) {
+        // TODO Do we need that ?
+        val policy = StrictMode.ThreadPolicy.Builder().permitAll().build()
+        StrictMode.setThreadPolicy(policy)
+
+        val avatarView = noticeView.findViewById<ImageView>(R.id.contributorProfile)
+        val avatarBitmap =
+          BitmapFactory.decodeStream(contributorAvatarUrl.openConnection().getInputStream())
+        avatarView.setImageBitmap(avatarBitmap)
+      }
+    }
+
+    liveNotice.observeForever(noticeObserver) // TODO we should link to a LifecycleOwner
+
+    noticeView
```

Comment:
Looks like a wip, but might be a language feature I'm not aware of.

Reply:
Yep, last line is the return value !